libreoffice-starter-extension-master/source/org/libreoffice/example/dialog/ActionTwoAndThree.java

```diff
@@ -0,0 +1,145 @@
+	private static com.sun.star.text.XTextViewCursor xTextViewCursor;
+	/** Contains all translation paragraphs */
+	private static String combined = "";
+
+	/**
+	 * @param xContext
+	 */
+	public ActionTwoAndThree (XComponentContext xContext) {
+		this.xContext = xContext;
+	}
+
+	/**
+	 * Append tranlation to the end of selected area.
+	 * Clean the variable that contains the translation.
+	 *
+	 * @throws Exception if getting translation while combining paragraphs failed
+	 */
+	public void appendAction() throws Exception {
+		combineTranslatedParagraphs();
+		append(combined);
+		combined = "";
+
+	}
+
+	/**
+	 * Moves cursour to the end of selection to insert translation there.
+	 *
+	 * @param translation	String containing translated text
+	 */
+	private void append(String translation) {
+		xTextViewCursor.collapseToEnd();
+		xTextViewCursor.setString(translation);
+	}
+
+	/**
+	 * If translated text is not empty, replace it with the translation.
+	 * Clean the variable that contains the translation.
+	 *
+	 * @throws Exception if getting translation while combining paragraphs failed
+	 */
+	public void replaceAction() throws Exception {
+		combineTranslatedParagraphs();
+		if(combined.length() > 1) {
+			combined = combined.substring(1); //remove unnecessary "\n" at the beginning
+			replace(combined);
+			combined = "";
+		}
+	}
+
+	/**
+	 * Replace selected text with it's translation
+	 *
+	 * @param translation	String containing all translated text
+	 */
+	private void replace(String translation) {
+		xTextViewCursor.setString(translation);
+	}
+
+	/**
+	 * Gets the user translatable input text
+	 *
+	 * @return	String that user wrote in dialog's translation box
+	 */
+	private String getSelectedText() {
+		com.sun.star.text.XTextDocument xTextDoc = DocumentHelper.getCurrentDocument(xContext);
+		com.sun.star.frame.XController xController = xTextDoc.getCurrentController();
+		com.sun.star.text.XTextViewCursorSupplier xTextViewCursorSupplier = DocumentHelper.getCursorSupplier(xController);
+		xTextViewCursor = xTextViewCursorSupplier.getViewCursor();
+
+		return xTextViewCursor.getString();
+	}
+
+	/**
+	 * To avoid errors from "\n" characters in translatable text,
+	 * each paragraph is translated seperately.
+	 * To make insertion of translated text as one step for user,
+	 * all translated paragraphs are combined in a single variable.
+	 *
+	 * @throws Exception	if getting translation failed
+	 */
+	private void combineTranslatedParagraphs() throws Exception {
+		String selectedText = getSelectedText();
+		String smt = TildeTranslatorImpl.getSystemID();
+		String clientID = TildeTranslatorImpl.getClientID();
+		String translation = null;
+		if(selectedText.length() > 0) {
+			String paragraphs[] = selectedText.split("\\r?\\n");
+			int paralength = paragraphs.length;
+			// translate each paragraph seperately
+			for(int i = 0; i != paralength; i++) {
+				translation = TranslateAPI.translate(clientID, smt, paragraphs[i]);
```

Comment:
No pašrakstītās kombinēšanas loģikas varētu atbrīvoties, tā vietā izmantojot String masīvu apvienošanas funkciju: `String.join("\n", paragraphs);`
`translate` izsaukums, gan būtu jāpamaina. Piemēram, varētu izmantot esošo source teksta masīvu, kuru tālāk izmantot tāpat nevajadzēs:
`paragraphs[i] = TranslateAPI.translate(clientID, smt, paragraphs[i])`
un beigās tad (space pievienošanas logika gan papildus jāveic):
`translation = String.join("\n", paragraphs);`

Šādas izmaiņas gan nozīmētu, ka jāpamaina funkcijas interfeis, lai varētu atgriezt string vērtību un jāatbrīvojas no statiskā `combine` klases mainīgā. Tas arī padarītu šo klasi thread drošu, kas ļautu izpildīt to vienlaicīgi vairākas reizes nesabojājot vienam otra tulkojumu. Varbūt šajā implementācija tāda situācija nevarētu notikt, bet kopumā tas būs tīrāks un uztveramāks klases dizains.

Reply:
Par papildus "\n" vai " " pievienošanu vajadzētu būt loģikai tieši `replaceAction` un `appendAction` metodēs